src/library_browser.js

```diff
@@ -766,57 +766,75 @@ var LibraryBrowser = {
 
-#if PTHREADS
-    if (ENVIRONMENT_IS_PTHREAD) {
-      err(`emscripten_async_load_script("${url}") failed, emscripten_async_load_script is currently not available in pthreads!`);
-      return onerror ? onerror() : undefined;
-    }
-#endif
-#if ASSERTIONS
-    assert(runDependencies === 0, 'async_load_script must be run when no other dependencies are active');
-#endif
-    {{{ runtimeKeepalivePush() }}}
+    // Creating a promise for script loading functionality
+    return new Promise((resolve, reject) => {
+    
+      #if PTHREADS
+      if (ENVIRONMENT_IS_PTHREAD) {
+        err(`emscripten_async_load_script("${url}") failed, emscripten_async_load_script is currently not available in pthreads!`);
+        return reject('emscripten_async_load_script is not available for pthreads');
+      }
+      #endif
+      
+      #if ASSERTIONS
+      assert(runDependencies === 0, 'async_load_script must be run when no other dependencies are active');
+      #endif
+      
+      {{{ runtimeKeepalivePush() }}}
 
-    var loadDone = () => {
-      {{{ runtimeKeepalivePop() }}}
-      if (onload) {
-        if (runDependencies > 0) {
-          dependenciesFulfilled = onload;
-        } else {
-          onload();
-        }
+      var loadDone = () => {
+        {{{ runtimeKeepalivePop() }}}
+        resolve();
       }
-    }
 
-    var loadError = () => {
-      {{{ runtimeKeepalivePop() }}}
-      onerror?.();
-    };
+      var loadError = () => {
+        {{{ runtimeKeepalivePop() }}}
+        reject();
+      };
 
-#if ENVIRONMENT_MAY_BE_NODE && DYNAMIC_EXECUTION
-    if (ENVIRONMENT_IS_NODE) {
-      readAsync(url, (data) => {
-        eval(data);
-        loadDone();
-      }, loadError, false);
-      return;
-    }
-#endif
+      #if ENVIRONMENT_MAY_BE_NODE && DYNAMIC_EXECUTION
+      if (ENVIRONMENT_IS_NODE) {
+        readAsync(url, (data) => {
+          eval(data);
+          loadDone();
+        }, loadError, false);
+        return;
+      }
+      #endif
+
+      var script = document.createElement('script');
+      script.onload = loadDone;
+      script.onerror = loadError;
+      script.src = url;
+      document.body.appendChild(script);
 
-    var script = document.createElement('script');
-    script.onload = loadDone;
-    script.onerror = loadError;
-    script.src = url;
-    document.body.appendChild(script);
+    });
   },
 
+
+  // Refactored the function to implement the promise based functionality and backward compatibility with callbacks
+  emscripten_async_load_script__deps: ['emscripten_async_load_script_promise', '$UTF8ToString'],
+  emscripten_async_load_script: function(url, onload, onerror)  {
+      
+      emscripten_async_load_script_promise(url).then(() => {
+        if (onload){
+          onload();
+        }
+      }).catch((error) => {
+        if (onerror) {
+          onerror();
+        } else {
+           err('emscripten_async_load_script error: ${error.message}');
+        }
+     });
+
+    },
```

Comment:
If this approach is to be generalized then perhaps this can be replaced with a helper function. e.g:


```
emscripten_async_load_script: function(url, onload, onerror) {
  promiseToCallbacks(emscripten_async_load_script_promise(url), onload, onerror);
}
```

However, I think it might be best to have two helpers: `promiseToCallbacks` and `callbacksToPromise`, and then use the one that makes sense for the underlying API.  I suggest we only use `promiseToCallbacks` for APIs where the underlying API is already promise based.

Reply:
defined and used the helper functions @sbc100